### DISCUSSION TITLE
How to start KMSCON with tuigreet + greetd?

### DISCUSSION THREAD

**Discussion by s1syph0s** [Aetf/kmscon#76] [Q&A]
Hi,

so I want to install [tuigreet](https://github.com/apognu/tuigreet) and [greetd](https://wiki.archlinux.org/title/Greetd) in my PC. Problem is, I am using default VT, and it doesn't look polished in comparison to the picture that was uploaded in [tuigreet](https://github.com/apognu/tuigreet)'s github. So I found this project to have a better quality rendering for fonts.

Now I have several questions:

According to [Arch's wiki](https://wiki.archlinux.org/title/KMSCON), I should start kmscon using service.

> To enable kmscon on tty1, [disable](https://wiki.archlinux.org/title/Disable) getty@tty1.service and [enable](https://wiki.archlinux.org/title/Enable) kmsconvt@tty1.service.

So I did that, and using my config:

```
# /etc/greetd/config.toml
[terminal]
# The VT to run the greeter on. Can be "next", "current" or a number
# designating the VT.
vt = 1

# The default session, also known as the greeter.
[default_session]

# `agreety` is the bundled agetty/login-lookalike. You can replace `/bin/sh`
# with whatever you want started, such as `sway`.
#command = "agreety --cmd /bin/sh"
command = "tuigreet -t --cmd Hyprland"

# The user to run the command as. The privileges this user must have depends
# on the greeter. A graphical greeter may for example require the user to be
# in the `video` group.
user = "greeter"
```

I should start my greetd using VT1. The problem is, when I reboot my PC, it shows logind using KMSCON (It should instead be showing tuigreet), and after logging in, I started another getty console..

Then I found this link:
https://todo.sr.ht/~kennylevinsen/greetd/36

According to this link, I need to configure the greetd command like this:
```
[default_session]
command = "kmscon -- tuigreet -t --cmd Hyprland"
```

But running this would give an error and prevents me from booting.

Did anyone have experience with this? What did I do wrong here?

**Comment by niten94** [comment]
I have tried looking on how to use greeters in kmscon, but this currently isn't possible without workarounds.

First, there are points in the post that may need to be clarified:

<details>

1. With usual configuration, the kmscon service cannot be started while greetd is running on the same VT. greetd sets the standard streams of commands it runs to `/dev/ttyN` unlike kmscon, since user space terminal emulators need to run commands on any available pseudoterminal in `/dev/pts` instead.

2. I think you misunderstood some programs that are run when using the default kmscon service. They don't need to be understood well in this case, but the description and order in which they're run are indicated below:
   - kmscon - runs getty as specified in service file
   - getty - usually sets TTY settings, prints `/etc/issue` and prompts username
   - login - reads password of user passed from getty and starts session with shell

   logind is a daemon in systemd that manages seats and sessions, so it's a lot different with `login`.

3. The greetd author didn't mention that the format of the `command` configuration key is exactly `kmscon -- command`.

</details>

To use greeters in kmscon, using a command like `kmscon -l --no-reset-env -- tuigreet -t` in greetd configuration and disabling kmscon services should work. However this will not work after logout if systemd is used, because permissions on `/dev/ttyN` are reset when display servers exit: systemd/systemd#10103

The greeter user needs to be added in groups with access to files below:
- `/dev/dri/cardN` - logind grants permissions to the current user, so adding to the associated group is only needed when systemd isn't used
   - `/dev/fbN` if DRM isn't available, but group is required even with systemd
- `/dev/input/*` - Display servers (Xorg server, Wayland compositors) open these files without group permissions, by taking a descriptor through `login1` D-Bus interface instead unlike kmscon
- `/dev/ttyN` - greetd and DMs (GDM, SDDM, etc.) don't set the owner of the file to the current user unlike `login`
   - The owner of the file is set if kmscon would call `TakeControl` method on `login1` D-Bus interface like display servers
   - Group read-write permission could be granted at greetd service startup

A workaround is also needed to address the points below:
1. kmscon restarts when the command passed to it exits
2. VT hangs or remains blank when greetd is stopped while kmscon is running
   - kmscon receives `SIGHUP` then exits due to the greetd session worker immediately stopping
   - logind seems to restore VT to text mode if a program that exits does use `TakeControl`/`ReleaseControl`
   - greetd service can be modified to run a program that enables text mode, but there's no command that sets this

I have created this repository with scripts that run greeter in kmscon only once and use sudo to set VT owner (may work with systemd): https://github.com/niten94/kmscon-in-greetd

**Reply by mock1328** [reply]
Thanks for research on this topic. I was trying to make it run for quite some time, but due to lack of knowledge I wasn't able to achieve any progress. Now to look at this, it is a much more complicated than it looked at first

**Reply by yheuhtozr** [reply]
@niten94 I appreciate this answer very much as it is exactly what I needed to know, but could you elaborate more?

Specifically, when:
```
$ id greeter
uid=994(greeter) gid=992(greeter) groups=992(greeter),3(tty),26(video),174(input)
```
and the command is:
```
kmscon -l --no-reset-env --no-switchvt -- tuigreet --cmd ...
```

it still emits following errors:
```
[0000.008992] ERROR: vt: cannot open tty /dev/tty1 (13): Permission denied (open_tty() in ../src/uterm_vt.c:289)
[0000.009626] ERROR: cannot create seat object on seat seat0: -13 (app_seat_new() in ../src/kmscon_main.c:221)
```

**Reply by niten94** [reply]
I do not know if you have changed `/dev/tty1` permissions but there are 2 specific possible reasons the error occurs:
1. On some distributions, `tty` group may have no read permission on `/dev/tty*` or the group of the files is `root`
2. systemd is used, so permissions are reset when a display server like X11 or Wayland compositors exited before kmscon starts
   - Modifying permissions with a method like running a command before the greetd service starts will not always work, since the owner of `/dev/ttyN` will be reset on logout while the service continues to run

I have slightly modified my comment, and added a link to a repository I just created which contains script files that implement a workaround to set the owner of `/dev/ttyN`.

**Reply by michael-oberpriller** [reply]
@niten94 Kind of unrelated, but are you saying that if kmscon would use TakeControl() on ttys it could open the tty without being root? I am interested in allowing kmscon to run as a non root user and opening ttys is one issue.

**Reply by niten94** [reply]
I'm sorry for the late reply. My comments contain related details but weren't clearly written at the time, so I'll also explain some points again.

> Kind of unrelated, but are you saying that if kmscon would use TakeControl() on ttys it could open the tty without being root?

Yes, but more specifically `TakeControl()` is used on the user's whole session. It sets the session controller to the calling process, and grants it access to all devices under the session including the TTY.

I think this is the official documentation of the session objects on D-Bus:
https://www.freedesktop.org/software/systemd/man/258/org.freedesktop.login1.html#Session%20Objects

> I am interested in allowing kmscon to run as a non root user and opening ttys is one issue.

I see. These may also be the only other files needed:
- `/dev/dri/cardN` (if using DRM backend) - opened directly (like `/dev/ttyN`)
- `/dev/input/*` - seems to be opened through `TakeDevice()`, which would probably require `libudev` to determine what arguments to pass instead of the path
- `/dev/fbN` (if using fbdev backend) - opened like `/dev/input/*` but not sure if it works with `login1`, and fbdev in general seems to be less prioritized than DRM

There are a lot of areas I'm not familiar especially with the actual implementation, so it might be good to look into the code of software like [wlroots](https://gitlab.freedesktop.org/wlroots/wlroots/) and [libseat](https://sr.ht/~kennylevinsen/seatd/). Unfortunately, I don't think I've seen any detailed explanations or examples myself.

**Comment by celet-ff-io** [comment]
You may try this, a brief summary of my solution :

```
command = tuigreet --cmd ' sudo kmscon --vt=1 --seats=seats0 --no-switchvt --login -- /bin/login -p -f $(whoami)'
```

And this is my actual NixOS config):

```nix
command = ''
${pkgs.tuigreet}/bin/tuigreet \
    --greeting '${config.tuigreet.greeting}' \
    --time \
    --remember \
    --asterisks \
    --theme 'border=magenta;text=cyan;prompt=green;time=red;action=blue;button=yellow;container=black;input=red' \
    --cmd 'sudo \
    ${pkgs.kmscon}/bin/kmscon --vt=1 --seats=seat0 --no-switchvt \
    --configdir ${configDir} \
    --login -- ${pkgs.shadow}/bin/login -p -f $(whoami) \
    '
'';
```

This works for me in my NixOS stable 25.11. I leaned those kmscon args from the [offical nixos module kmscon](https://github.com/NixOS/nixpkgs/blob/nixos-25.11/nixos/modules/services/ttys/kmscon.nix). What this solution could not do is run kmscon succesfully by a user not in the "wheel" group.

What the command does: `tuigreet` login in a user (have to be in "wheel" in my solution), the user uses `sudo` to run kmscon succesfully, and login itself without password input with `login -p -f $(whoami)`. 

I found this solution because it seems only a user with root privileges can run kmscon successfully. And kmscon defaultly uses `login` as the command to run after you enter the kmscon screen. I tried to use `$SHELL` instead and it runs the shell as root, because I used `sudo` to run `kmscon`. So I switch to `login -p -f $(whoami)`, which login in without password as, the current user before `sudo` works.
